aten/src/ATen/DeviceGuard.h

```diff
@@ -6,131 +6,180 @@
 #include <ATen/detail/CUDAHooksInterface.h>
 #include <c10/util/Exception.h>
 #include "c10/util/Optional.h"
+#include <c10/detail/DeviceGuardImplInterface.h>
 
 #include <cstddef>
 
 namespace at {
+
 /// RAII guard that sets a certain default device in its constructor, and
-/// changes it back to the device that was originally active upon destruction.
+/// changes it back to the device (for that device type) that was originally
+/// active upon destruction.
 ///
-/// The device is always reset to the one that was active at the time of
-/// construction of the guard. Even if you `set_device` after construction, the
-/// destructor will still reset the device to the one that was active at
-/// construction time.
-struct DeviceGuard {
-  /// Default constructor, does nothing.
-  DeviceGuard() = default;
-
+/// If a DeviceGuard is constructed without specifying a device type (this can
+/// occur if you, e.g., pass a nullopt to the constructor), it behaves as if it
+/// were a no-op "CPU" guard; e.g., current_device() reports that the current
+/// device is kCPU.  This is different from passing Device(kCUDA, -1), which
+/// says to use the current CUDA device; in this case, we will correctly query
+/// what the current CUDA device is, won't change it, but WILL reset it
+/// at the end of DeviceGuard.
+class DeviceGuard {
+public:
   /// Set the current device to the passed Device.
   explicit DeviceGuard(Device device) {
-    set_device(device);
+    init_device(device);
   }
 
-  explicit DeviceGuard(c10::optional<Device> device_opt) {
+  /// Set the current device to the passed Device, if not nullopt;
+  /// otherwise do nothing.
+  explicit DeviceGuard(optional<Device> device_opt) {
     if (device_opt.has_value()) {
-      set_device(device_opt.value());
+      init_device(device_opt.value());
     }
   }
 
   /// Sets the current device to the device on which the given tensor is located.
   explicit DeviceGuard(const Tensor& tensor) {
-    set_device_from(tensor);
+    init_device_from(tensor);
   }
 
   /// Sets the current device to the device on which the first tensor in the list is
   /// located. If the list is empty, does nothing.
   explicit DeviceGuard(const TensorList& tensors) {
     if (!tensors.empty()) {
-      set_device_from(tensors.front());
+      init_device_from(tensors.front());
     }
   }
 
+  /// A constructor for testing; permits explicitly passing in the
+  /// DeviceGuardImpl.
+  explicit DeviceGuard(Device device, const detail::DeviceGuardImplInterface* impl) {
+    init_device(device, impl);
+  }
+
   /// Copy is disallowed.
   DeviceGuard(const DeviceGuard&) = delete;
   DeviceGuard& operator=(const DeviceGuard&) = delete;
 
-  /// Move-constructs this `DeviceGuard` from another `DeviceGuard`. The
-  /// moved-from `DeviceGuard` is modified such that its destruction has no
-  /// effect (does not reset the device).
+  /// Move-constructs this `DeviceGuard` from another `DeviceGuard`.
```

Comment:
where do we use move constructors actually? If all we need is "diffusing" the DeviceGuard - having guard.dismiss() seems like more explicit API.

Reply:
You mean "defusing" :)

The move constructor is occasionally useful. It's currently used here:

```
  // NB: We don't need to hold the device guard when calling from TH
  c10::optional<cuda::CUDAGuard> guard(
    device_guard ?
      make_optional(cuda::CUDAGuard(self->storage().device().index())) :
      nullopt);
```

This won't compile without the move constructor. Also, this is not equivalent to the optional constructor on CUDAGuard, since that constructor will still query the current device and restore it on exit.

Reply:
Haha, "diffusing" - too much Phabricator (should it be "diphphusing?")

Interesting, took me a while to understand that code :)

Would `optional::emplace` work? It doesn't require move constructor it seems: https://godbolt.org/z/PYP18V

Reply:
Yeah, I think it will.

BTW, I don't think we should support dismissal, it still has the same problem as move construction where it's hard to make it do something reasonable if there's an intervening device guard between the dismiss site, and the original device guard.

Reply:
OK, so there's another use of the move constructor which is harder to remove in C++11. It looks like this:

```
  at::cuda::CUDAGuard createStreamGuard() {
    at::cuda::CUDAGuard guard;
    for (const auto& stream : streams_) {
      guard.set_stream(stream);
    }
    return guard;
  }
```

NRVO will often eliminate the move constructor call here, but it is not mandatory (even in C++17) and so we are required to show that there is a move constructor.

Reply:
To make matters worse, I can't even replace this with `return CUDAGuard(streams_)` (adding a new constructor that takes a list of streams) because in C++11 the copy elision is not mandatory (yet).

Reply:
Well, I resolved it just by inlining everything.

Reply:
ok, convinced me. I think move constructor is ok then if you explicitly write that reordering is  UB (which you did). I just checked folly::Guard and it's move-constructible too

Reply:
The inlining wasn't too bad, once I added the ArrayRef<CUDAStream> constructor. And it let me eliminate "uninitialized guard" state, which means simpler code, yay.